File: 2024/day24-24/day24-24.py

# read command-line parameters and based on that read the input file
from collections import defaultdict
from itertools import product
import random
import sys
runtype = sys.argv[1]
runpart = int(sys.argv[2])
if len(sys.argv) > 3:
    runs = int(sys.argv[3])
else:
    runs = 1

text_file = open(f"day24-{runtype}.txt", "r")

lines = [line.strip() for line in text_file.readlines()]
empty = lines.index('')
inputs = { line[0:3]: int(line[4:]) for line in lines[:empty] }

gates = {}
for line in lines[empty+1:]:
    f, t = line.split(' -> ')
    if 'OR' in f:
        f = f.replace(' OR', ' OR ')
    i1 = f[0:3]
    op = f[4:7]
    i2 = f[8:11]
    gates[t] = (i1, op, i2)


def process(wires):
    global gates
    active = {out:op for (out,op) in gates.items() if op[0] in wires and op[2] in wires}

    output = {}
    for out, gate in active.items():
        i1 = wires[gate[0]]
        i2 = wires[gate[2]]
        if gate[1] == 'AND':
            output[out] = i1 & i2
        elif gate[1] == 'OR ':
            output[out] = i1 | i2
        elif gate[1] == 'XOR':
            output[out] = i1 ^ i2
        else:
            raise Exception("Unknown operator")
    return output

def get_decimal_value(wires):
    v = 0
    for i in range(len(wires)-1, -1, -1):
        z_str = '{:02d}'.format(i)
        for name, value in wires.items():
            if name.endswith(z_str):
                v |= value << i
    return v

def process_until_target_are_set(wires, target):
    global gates
    target_values = {t: None for t in target}
    loops = 0
    while any([v == None for v in target_values.values()]) and loops < 100:
        output = process(wires)
        wires.update(output)
        for t in target:
            if t in output:
                target_values[t] = wires[t]
        loops += 1
        
    if loops == 100:
        return None
    return target_values

def part1():
    global gates
    z_gates = {k: v for k, v in gates.items() if k.startswith('z')}
    z_values = process_until_target_are_set(inputs, z_gates.keys())
    res1 = get_decimal_value(z_values)
    return res1

# Just reconstruct the full adder from the resulting bit gate
# All cases that are wrong will return the output wire that is wrong
def check_back(bit):
    global gates
    gate_name = f"z{bit:02d}"
    if gate_name not in gates: return False, gate_name, "not in gates"
    xor_gate = gates[gate_name]
    if xor_gate[1] != "XOR": return False, gate_name, "not XOR"
    
    in_wires = {xor_gate[0], xor_gate[2]}
    in_gates = {gn:gates[gn] for gn in in_wires}

    xor_gate_name = {gn for gn, gate in in_gates.items() if gate[1] == "XOR"}
    if len(xor_gate_name) == 0: 
        if bit == 1:
            return False, {gn for gn, gate in in_gates.items() if gate[1] != "AND"}.pop(), "initial bit and is wrong"
        else:
            return False, {gn for gn, gate in in_gates.items() if gate[1] != "OR "}.pop(), "carry is wrong"
    
    x_wire = f"x{bit:02d}"
    y_wire = f"y{bit:02d}"
    
    xor_gate_inputs = {(gate[0], gate[2]) for gn, gate in in_gates.items() if gate[1] == "XOR"}
    if len(xor_gate_inputs) != 1:
        incorrect_in_wires = [xgi for xgi in xor_gate_inputs if {xgi[0], xgi[1]} != {x_wire, y_wire}].pop()
        incorrect_in_wires = {incorrect_in_wires[0], incorrect_in_wires[1]}
        return False, [n for n, g in gates.items() if {g[0], g[2]} == incorrect_in_wires].pop(), "carry is wrong"

    if bit == 1:
        and_gate = {gate for gate in in_gates.values() if gate[1] == "AND"}.pop()
        if {and_gate[0], and_gate[2]} != {f"x00", f"y00"}:
            return False, {gn for gn, gate in in_gates.items() if gate[1] != "AND"}.pop(), "initial carry is wrong"
    else:
        or_gate = {gate for _, gate in in_gates.items() if gate[1] == "OR "}.pop()
        
        in_or_gate =  {(or_gate[0], gates[or_gate[0]]), (or_gate[2],gates[or_gate[2]])}
        in_or_gate_not_and = {n for (n,g) in in_or_gate if g[1] != "AND"}
        if len(in_or_gate_not_and) != 0:
            return False, in_or_gate_not_and.pop(), "and before carry is wrong"
    return True, None, "all good"


def part2():
    return ",".join(sorted(list({out for (_,out,_) in [check_back(i) for i in range(1, 45)]} - {None})))

if runpart == 1 or runpart == 0:
    for run in range(runs):
        resp1 = part1()
    print("Part 1: {}".format(resp1))

if runpart == 2 or runpart == 0:
    for run in range(runs):
        resp2 = part2()
    print("Part 2: {}".format(resp2))
    

File: 2024_rust/src/day24_24.rs
use super::tools;
use std::time::Instant;
use std::collections::{HashMap, HashSet};
use itertools::Itertools;

fn process(wires: &HashMap<String, i64>, gates: &HashMap<String, (String, String, String)>) -> HashMap<String, i64> {
    let mut output: HashMap<String, i64> = HashMap::new();
    
    let active = gates.iter().filter(|(_, (in1, _, in2))| {
        wires.contains_key(in1) && wires.contains_key(in2)
    }).collect::<HashMap<&String, &(String, String, String)>>();
    
    for (out, (in1, op, in2)) in active {
        let in1 = wires.get(in1).unwrap();
        let in2 = wires.get(in2).unwrap();
        let result = match op.as_str() {
            "AND" => in1 & in2,
            "OR " => in1 | in2,
            "XOR" => in1 ^ in2,
            _ => panic!("Unknown operator"),
        };
        output.insert(out.to_string(), result);
    }
    output
}

fn get_decimal_value(wires: HashMap<String, i64>) -> i64 {
    let mut v = 0;
    let mut bit = wires.len() as i64 - 1;
    while bit >= 0 {
        let z_str = format!("z{:02}", bit);
        let value = wires[&z_str];
        v |= value << bit;
        bit -= 1;
    }
    v
}

fn process_until_target_are_set(wires: HashMap<String, i64>, gates: &HashMap<String, (String, String, String)>, target: Vec<String>) -> HashMap<String, i64> {
    let mut target_values = HashMap::new();

    let mut updated_wires = wires.clone();

    while target_values.len() < target.len(){
        let mut output = process(&updated_wires, gates);
        for (k, v) in output.drain() {
            *updated_wires.entry(k.clone()).or_insert(0) = v;
            if target.contains(&k) {
                target_values.insert(k, v);
            }
        }
    }
    return target_values;
}

fn check_back(bit: usize, gates: &HashMap<String, (String, String, String)>) -> Result<String, String> {
    let gate_name = format!("z{:02}", bit);
    if !gates.contains_key(&gate_name) {
        return Err(gate_name);
    }
    let xor_gate = gates.get(&gate_name).unwrap();
    if xor_gate.1 != "XOR" {
        return Err(gate_name);
    }

    let in_wires: HashSet<String> = HashSet::from_iter(vec![&xor_gate.0, &xor_gate.2].iter().map(|x| x.to_string()));
    let in_gates = gates.iter().filter(|&(out, _)| in_wires.contains(&(out.clone()))).collect::<HashMap<&String, &(String, String, String)>>();

    let xor_gate_names = in_gates
        .iter()
        .filter(|(_, (_, op, _))| *op == "XOR")
        .map(|(out, _)| out.to_string())
        .collect::<Vec<String>>();

    let not_and_gate_names = in_gates
        .iter()
        .filter(|(_, (_, op, _))| *op != "AND")
        .map(|(out, _)| out.to_string())
        .collect::<Vec<String>>();

    let not_or_gate_names = in_gates
        .iter()
        .filter(|(_, (_, op, _))| *op != "OR ")
        .map(|(out, _)| out.to_string())
        .collect::<Vec<String>>();

    if xor_gate_names.len() == 0 {
        if bit == 1 {
            return Err(not_and_gate_names[0].clone());
        } else {
            return Err(not_or_gate_names[0].clone());
        }
    }

    let x_wire = format!("x{:02}", bit);
    let y_wire = format!("y{:02}", bit);

    let xor_gate_inputs = in_gates
        .iter()
        .filter(|(_, (_, op, _))| *op == "XOR")
        .map(|(_, (in1, _, in2))| (in1.to_string(), in2.to_string()))
        .collect::<HashSet<(String, String)>>();

    if xor_gate_inputs.len() != 1 {
        let incorrect_in_wires = xor_gate_inputs
            .iter()
            .filter(|(in1, in2)| !(in1 == &x_wire && in2 == &y_wire || in1 == &y_wire && in2 == &x_wire ))
            .fold(HashSet::new(), |mut acc, (in1, in2)| {
                acc.insert(in1.clone());
                acc.insert(in2.clone());
                acc
            });
        let incorrect_gate = &gates
            .iter()
            .filter(|(_, (in1, _, in2))| incorrect_in_wires.contains(in1) && incorrect_in_wires.contains(in2))
            .map(|(out, _)| out.to_string())
            .collect::<Vec<String>>()[0].clone();
        return Err(incorrect_gate.to_string());
    }

    if bit == 1 {
        let and_gate: HashSet<String> = in_gates
            .iter()
            .filter(|(_,(_,op,_))| *op == "AND")
            .fold(HashSet::new(), |mut acc, (_, gate)| {
                acc.insert(gate.0.to_string());
                acc.insert(gate.2.to_string());
                acc
            });
        
        let initial_gate = HashSet::from_iter(
            vec!["x00", "y00"]
            .iter()
            .map(|x| x.to_string())
        );

        if initial_gate != and_gate {
            return Err(not_and_gate_names[0].clone());
        }

    } else {
        let or_gate = &in_gates
            .iter()
            .filter(|(_,(_,op,_))| *op == "OR ")
            .map(|(_, gate)| *gate)
            .collect::<Vec<&(String, String, String)>>()[0];

        let mut in_or_gate: HashSet<(String, (String, String, String))> = HashSet::new();
        in_or_gate.insert((or_gate.0.clone(), gates.get(&or_gate.0).unwrap().clone()));
        in_or_gate.insert((or_gate.2.clone(), gates.get(&or_gate.2).unwrap().clone()));
        let in_or_gate_not_and = in_or_gate
            .iter()
            .filter(|(_,(_,op,_))| *op != "AND")
            .map(|(out, _)| out.to_string())
            .collect::<Vec<String>>();
        if in_or_gate_not_and.len() != 0 {
            return Err(in_or_gate_not_and[0].clone());
        }
    }

    return Ok("".to_string());
}

#[allow(dead_code)]
pub fn run(real: bool, print_result: bool) -> (u128, u128, u128) {
    let start0 = Instant::now();

    let input_file: &str = if !real {
        "./input/day24-test.txt"
    } else {
        "./input/day24-real.txt"
    };

    let input = tools::get_input(String::from(input_file));
    let empty = input.iter().enumerate().filter(|(_, x)| *x == "").map(|(i, _)| i).collect::<Vec<usize>>()[0];
    let inputs = input[..empty].iter().fold(HashMap::new(), |mut acc, x| {
        acc.insert(x[0..3].to_string(), x[5..].to_string().parse::<i64>().unwrap());
        acc
    });

    let gates = input[empty+1..]
        .iter()
        .fold(HashMap::new(), |mut acc, line| {
            let inps = line.split(" -> ").collect::<Vec<&str>>();
            let out = inps[1].to_string();
            let inps = inps[0].split(" ").collect::<Vec<&str>>();
            let mut op = inps[1].to_string();
            if op == "OR" { op.push(' '); }
            let in1 = inps[0].to_string();
            let in2 = inps[2].to_string();
            acc.insert(out, (in1, op, in2));
            acc
        });

    let after0 = Instant::now();

    let start1 = Instant::now();

    let z_gates = gates
        .iter()
        .filter(|(out, _)| out.starts_with("z"))
        .map(|(out, _)| out.to_string())
        .collect::<Vec<String>>();
    let z_values = process_until_target_are_set(inputs, &gates, z_gates);
    let res1 = get_decimal_value(z_values);

    let after1 = Instant::now();
    if print_result {
        println!("Part 1: {}", res1);
    }

    let start2 = Instant::now();

    let res2 = (1..45)
        .map(|bit| check_back(bit, &gates) )
        .filter(|x| x.is_err())
        .map(|x| match x { Ok(gate) => {gate}, Err(gate) => gate })
        .collect::<Vec<String>>()
        .iter()
        .sorted()
        .join(",");

    let after2 = Instant::now();
    if print_result {
        println!("Part 2: {}", res2);
    }

    (
        after0.duration_since(start0).as_nanos(),
        after1.duration_since(start1).as_nanos(),
        after2.duration_since(start2).as_nanos(),
    )
}
